Question Tile MVP

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Adds ability to layout questions horizontally by assigning a 'n-per-row' appearance attribute to questions where "n" is any integer (i.e. "3-per-row"). The question tile width will reduce accordingly and sequential tiles will be displayed such that the sum of the widths of each question tile in the row will be <=12 (12 is from the grid defined in Bootstrap)

https://github.com/dimagi/commcare-hq/assets/88759246/4ce26f0b-1a2f-4908-99b1-1bb5232816e5

![Screen Shot 2023-08-08 at 11 20 16 AM](https://github.com/dimagi/commcare-hq/assets/88759246/53071a78-112d-4b68-9ab0-c2c3b22b51a0)


## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Jira Ticket: https://dimagi-dev.atlassian.net/browse/USH-3418
Design Doc: https://docs.google.com/document/d/1WujrvTByNZaWvauR4e3Md-LLxt0qUpd451iB0PyHoo4/edit#heading=h.ipgou4bu0qoj

For defining the widths of the question tile, we utilize Bootstraps built in [grid](https://getbootstrap.com/docs/5.0/layout/grid/) which defines 12 columns. 

To display sequential question tiles with a combined "12" width on the same row and have following question tiles that do not fit overflow into the next row, each "12" width group is nested in a `<div class="row">`. The class `row` is leveraged from bootstrap. 
On the view model side, I added a `GroupedQuestionTileRow` where each instance contains questions that would all belong in a row. How the form_ui is is built is that a form, question, group, or repeat-juncture objects are defined in a json where objects would be nested in a `children`. I.e. the form contains a group child and the group contains questions children. To create the `GroupedQuestionTileRow` I recursively iterated through the json and grouped questions contained in `children` accordingly (The `GroupedQuestionTileRow` could contain only one question if the question is intended to take up the entire "12" width). As a result, this `GroupedQuestionTileRow` layer is added so a question will always be nested within a GroupedQuestionTileRow.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
None
## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Local testing. This change does impact all web app forms.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Tests already exist that entries in form are in the order as expected for the "Form" object. 
Added tests to check questions are grouped correctly when nested within the children of a "Form" Object

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Will pass through QA to check layout of forms with various question types and interactions.
https://dimagi-dev.atlassian.net/browse/QA-5472

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
feat: get and process new "3-per-row" appearance attribute

Jtang-1 committed:
feat: assign widths based on appearance attribute

Jtang-1 committed:
feat: apply widths to questions template

Jtang-1 committed:
refactor: extract function matching pattern to style string

Jtang-1 committed:
refactor: extract calculation of column widths for questiontile

Jtang-1 committed:
create view model for questions that will be grouped in the same row to be displayed horizontally

Jtang-1 committed:
modifies json by recursively grouping questions within nested containers

Jtang-1 committed:
bug: handle cases where resulting col width is a float

Jtang-1 committed:
bug: access questions within groupedQuestionTileRow layer

Jtang-1 committed:
linting

nospame commented on 12:
```diff
@@ -0,0 +1 @@
+Saved working directory and index state On master: grouped_mvp_actual only group those with width
```
Not sure what this is about; just a weird git artifact?

Jtang-1 replied:
I have no idea how that happened, but removed it [bda751e](https://github.com/dimagi/commcare-hq/pull/33333/commits/bda751ed3320f2e98a14209e5e977803a9ec4c36)

nospame reviewed (commented):
Sorry for so many comments! I think this looks good overall and really cool to see this take shape, and seems like a lot of little pieces to get working all together.

Jtang-1 committed:
remove accidentaly added file

Jtang-1 committed:
bug: handles GroupedQuestionTileRow required by iterating children

group nested in group previously tried to call an observable
that didn't exist for groups.

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -250,6 +255,49 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+                const questionTileWidth = Question.calculateColumnWidthForPerRowStyle(child.style);
+                usedWidth += questionTileWidth;
+
+                if (usedWidth > constants.GRID_COLUMNS) {
```
I think it should be `>`. Bootstrap column goes up to 12 (`constants.GRID_COLUMNS`). So 3 items with 4 column widths (12 total width) would be able to fit in one row. With `>=`, the 3rd item would be moved to the next row.

Jtang-1 committed:
Merge branch 'master' into jt/question_tile_mvp_actual_v0

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -111,6 +111,20 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
     }
 
+    function getMatchingStyles(pattern, styleStr) {
+        let retVal = [];
```
`matchingStyles` might be a better name.

Jtang-1 replied:
My preference is for names of non-boolean functions to be verbs so I would prefer it as is over `matchingStyles`

MartinRiese replied:
It's just that retVal does not tell you anything about the value stored in it.

Jtang-1 replied:
Oh my bad, I thought you were referring to the function name. Yes, I agree with you about `retVal`[1c8f778](https://github.com/dimagi/commcare-hq/pull/33333/commits/1c8f778e0a8c6054413f447d3eaed95cb97940b1)

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -111,6 +111,20 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+        if (styleStr) {
+            let styles = styleStr.split(' ');
+            styles.forEach(function (style) {
+                if ((pattern instanceof RegExp && style.match(pattern))
```
I think it would be enough to check the type of `pattern` once at the beginning of this function. And just return an empty list then.

Jtang-1 replied:
hm, i'm not following what you mean. I think the pattern type check determines how to match the pattern, so i think it needs to be done like how it is
```
if ((pattern instanceof RegExp && style.match(pattern))
      || (typeof pattern === "string" && pattern === style)
```

MartinRiese replied:
Oh. I misread the if statement.

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -250,6 +255,49 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+            usedWidth = 0;
+        }
+
+        for (const child of json.children) {
```
I am surprised it works with a const

Jtang-1 replied:
good point, [7f184ba](https://github.com/dimagi/commcare-hq/pull/33333/commits/7f184baf3b9f85716d7e69641bc7c4dbe81fb653)

I think generally const is ok (this is how[ mdn shows it](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)) but I am modifying child so `let` is probably more appropriate

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -738,15 +738,23 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
-        this.controlWidth = numPerRow !== null ? constants.FULL_WIDTH : constants.CONTROL_WIDTH;
-        this.labelWidth = numPerRow !== null ? constants.FULL_WIDTH : constants.LABEL_WIDTH;
-        this.questionTileWidth = numPerRow !== null ? `col-sm-${constants.GRID_COLUMNS / numPerRow}` : constants.FULL_WIDTH;
+        this.controlWidth = columnWidth === constants.GRID_COLUMNS ? constants.CONTROL_WIDTH : constants.FULL_WIDTH;
+        this.labelWidth = columnWidth === constants.GRID_COLUMNS ? constants.LABEL_WIDTH : constants.FULL_WIDTH;
+        this.questionTileWidth = columnWidth === constants.GRID_COLUMNS ? constants.FULL_WIDTH : `col-sm-${columnWidth}` ;
```
I agree

Jtang-1 replied:
Agreed, longer but more readable [7f184ba](https://github.com/dimagi/commcare-hq/pull/33333/commits/7f184baf3b9f85716d7e69641bc7c4dbe81fb653)

Jtang-1 committed:
refactor: extract pushing to current group

Jtang-1 committed:
bug: adds key to groupedQuestions

keys are used during ko.mapping such that the entire element isn't entirely replaced

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -735,6 +809,24 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+    Question.prototype.setWidths = function () {
+        const columnWidth = Question.calculateColumnWidthForPerRowStyle(this.style)
+
+        this.controlWidth = columnWidth === constants.GRID_COLUMNS ? constants.CONTROL_WIDTH : constants.FULL_WIDTH;
```
I don't think so since it's being called in the constructor, so `this` would point to the instance. But i know there's some weird thing with scope of `this`. What was the problem you ran into recently?

nospame replied:
I was setting up a util function as a screen resize listener that I wanted to run a callback function in the place it was called from in the same `this` context. https://github.com/dimagi/commcare-hq/pull/33330/commits/e00ddd200b6eee173a37128f0829bef731408430 Arrow functions ended up saving me a lot of grief, since they keep the same context.

Jtang-1 replied:
oh interesting... the callback seems to make it tricky, but it shouldn't be a problem with how it's used here.

Jtang-1 committed:
style changes

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -738,15 +738,23 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
     };
 
+    Question.calculateColumnWidthForPerRowStyle= function(style) {
+        let styleStr = (style) ? ko.utils.unwrapObservable(style.raw) : null;
```
yes, [7f184ba](https://github.com/dimagi/commcare-hq/pull/33333/commits/7f184baf3b9f85716d7e69641bc7c4dbe81fb653)

Jtang-1 committed:
style: add comments

Jtang-1 committed:
test: update tests to account for new GroupedQuestionTileRow parent layer of questions

Jtang-1 committed:
rename variable

Jtang-1 committed:
fix typo

Jtang-1 committed:
test: tests that questions are grouped according to width defined by style

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -282,13 +282,18 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
                 currentGroup = {
                     type: constants.GROUPED_QUESTION_TILE_ROW_TYPE,
                     children: [],
+                    ix: null,
```
What is `ix`?

Jtang-1 replied:
For questions, group, and repeat groups, the ix indexes the order. So form in this order would have the associated ix
1. Question (ix = "0")
2. Question (ix = "1")
3. Group (ix ="2")
3.1 Question nested in group (ix="2,0")
3.2 Question nested in group (ix = "2,1")

The way i'm using it here isn't really following this convention, since this "Grouped question tile row type" is sort of imaginary. I defined ix to be the concatenation of ix contained in the group. So in the example, Questions 1 and 2 would be grouped, and create a "grouped question tile row type" with ix = "(0),(1)".

The purpose I added it is that for [knockout mapping plugin](https://knockoutjs.com/documentation/plugins-mapping.html), we use this ix was a "key" to identify objects to see if they need to be replaced or just updated

nospame replied:
That's pretty cool, I didn't realize we could/do use knockout mapping that way. I think your usage makes sense with the different types of nesting that already exist. Thanks for the explanation!

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -649,7 +649,8 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
         Container.call(self, json);
 
-        self.groupRowChildrenRequired = ko.computed(function () {
+        self.required = ko.observable(0)
```
Good point - The data returned from formplayer for the other model types ("sub-group, question, and repeat-juncture") use `0` and `1` to as values for `required` so I just followed the convention.

Jtang-1 committed:
bug: when usedWidth resets to 0, it adds back width of current child

Jtang-1 committed:
lint

Jtang-1 committed:
feat: renders Bootstrap as "small" layout when printing

Bootstrap is written with mobile-first pattern so printing with paper size of
type "legal" results in normally using the "xs" view

nospame reviewed (commented):
Is it possible to do something like this answer to the bootstrap printing problem? https://stackoverflow.com/a/42414289
Seems cleaner but could maybe have unexpected effects on printing other parts of web apps (but so could this, I think)

nospame commented:
> @nospame Gave that a try and it doesn't seem to work. I think that just changes the default paper size (and restricts to print on only A4 since I don't have the option to change paper size when printing)

Was sort of wondering if it would be possible to use part of that rule it without the paper size, just to force a "larger" (or in some cases smaller) resolution for printing, something like:
```  
@media print {
  html, body {
    width: 1024px;
  }
}
```
But maybe the problem is in bootstrap's own `@media print` rules and that doesn't have any effect.

Jtang-1 commented:
@nospame This guy gave a pretty thorough explanation https://stackoverflow.com/a/51209031
From his answer: "No it won't work because media queries work on viewport and not on width of html or body". I also gave it a try and the outcome matches what he said.

nospame commented:
> @nospame This guy gave a pretty thorough explanation https://stackoverflow.com/a/51209031 From his answer: "No it won't work because media queries work on viewport and not on width of html or body". I also gave it a try and the outcome matches what he said.

That's... a mess. Finally found something I think Bootstrap is objectively wrong on (mobile-first display is great, but how often are you trying to print on something the size of a notecard?) Your solution is totally understandable.

Jtang-1 committed:
lint

Jtang-1 commented:
Failing javascript lint - reported failures don't seem to be right i.e reporting missing semicolons where there are already semicolons

snopoke commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js:
```diff
@@ -75,13 +75,48 @@ hqDefine("cloudcare/js/form_entry/spec/form_ui_spec", function () {
+                ix: "1",
+            });
+            g0.children[0].children[0].style = styleObj;
+            g0.children[0].children[0].style = styleObj;
```
should this be `g0.children[0].children[1].style = styleObj;`?

Jtang-1 replied:
yes https://github.com/dimagi/commcare-hq/pull/33389
